maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,320 @@
+
+    def _postpone_vm_requirement(self, postpone_type: PostponeType, vm_id: int, remaining_buffer_time: int):
+        """Postpone VM requirement."""
+
+        if remaining_buffer_time >= self._delay_duration:
+            if postpone_type == PostponeType.Resource:
+                self._total_latency.latency_due_to_resource += self._delay_duration
+            elif postpone_type == PostponeType.Agent:
+                self._total_latency.latency_due_to_agent += self._delay_duration
+
+            postpone_payload = self._pending_vm_req_payload[vm_id]
+            postpone_payload.remaining_buffer_time -= self._delay_duration
+            postpone_event = self._event_buffer.gen_cascade_event(
+                tick=self._tick + self._delay_duration,
+                event_type=Events.REQUIREMENTS,
+                payload=postpone_payload
+            )
+            self._event_buffer.insert_event(event=postpone_event)
+        else:
+            # Fail
+            # Pop out VM requirement payload.
+            self._pending_vm_req_payload.pop(vm_id)
+            # Add failed requirements.
+            self._failed_requirements += 1
+
+    def _get_valid_pm(self, vm_req_cpu: int) -> List[ValidPhysicalMachine]:
+        """Check all valid PMs.
+
+        There are three situations:
+            1. All PMs are empty.
+            2. Some PMs are empty but others are not.
+            3. All PMs are not empty.
+        Situation 1: Return the first empty PM info.
+        Situation 2: Return all PMs with enough resources but not empty and plus the first empty PM.
+        Situation 3: Return all PMs with enough resources but not empty.
+
+        Args: vm_req_cpu (int): The CPU resource of the VM requirement.
+        """
+        # NOTE: Should we implement this logic inside the action scope?
+        # TODO: In oversubscribable scenario, we should consider more situations, like
+        #       the PM type (oversubscribable and non-oversubscribable).
+
+        valid_pm_list = []
+        for pm in self._machines:
+            if pm.req_cpu == 0:
+                valid_pm_list.append(
+                    ValidPhysicalMachine(pm_id=pm.id, remaining_cpu=pm.cap_cpu, remaining_mem=pm.cap_mem)
+                )
+                break
+            elif pm.req_cpu > 0 and (pm.cap_cpu - pm.req_cpu) >= vm_req_cpu:
+                valid_pm_list.append(ValidPhysicalMachine(
+                    pm_id=pm.id,
+                    remaining_cpu=pm.cap_cpu - pm.req_cpu,
+                    remaining_mem=pm.cap_mem - pm.req_mem
+                ))
+        return valid_pm_list
+
+    def _on_vm_required(self, vm_requirement_event: CascadeEvent):
+        """Callback when there is a VM requirement generated."""
+        # Get VM data from payload.
+        payload: VmRequirementPayload = vm_requirement_event.payload
+
+        vm_req: VirtualMachine = payload.vm_req
```

Comment:
payload.vm_size or payload.vm_config?

Reply:
I agree that ``vm_req`` is not accurate enough for this.

But inside the current ``vm_req``, there is all the information of this VM (not only the requirement for CPU, mem, ... but also the future utilization series). So only ``vm_size`` or ``vm_config`` may also not good for this.

We need some more suitable naming or should we separate this infomation apart here?